# APS106 Design Problem 6
## Stock Portfolio - Problem Background

On Dec. 20, 1966, the New York Stock Exchange (NYSE) switched to a fully automated computerized trading system. A few years later, in 1971, Nasdaq launched as the world's first fully electronic stock market.

Today, the Nasdaq's digital marketplaces handle over 6 billion stock purchases, representing over $200 billion, each day!

This is a great example of how software can be used to model and emulate real-world objects and processes ([and a reminder of the real-world importance of testing and debugging code!](https://www.reuters.com/article/us-nasdaq-halt-glitch-idUSBRE97S11420130829)). In this design problem, you will practice file I/O and object-oriented programming to make your own stock market and stock portfolio.


## Part 1 - Read stock price data from csv file

### Define the Problem
In the first part of our problem, we want to read historical stock price data from a file and store it within variables in our program.

As an example, we'll use the csv file `netflix_stock_data.csv` which contains the daily Netflix stock prices between Oct. 26 and Nov. 18, 2022 (alternatively, you can download a csv file of Netflix stock prices for the last month here: https://www.nasdaq.com/market-activity/stocks/nflx/historical).

The first column in the csv file contains a date and the second column contains the price of a single stock on that date. We'll ignore the other columns for this problem.

One way to store this historical price information in our program is using a **dictionary**. We will use the dates (strings) as **keys** and the corresponding prices (floats) as **values**.

### Define Test Cases
We can use the file of Netflix stock prices as a test case. Some things that we will want to consider when determining if our solution is correct:
* Are the keys in the dictionary strings? Are they all valid dates?
* Are the values in the dictionary floats?
* Are the prices in the dictionary paired with the correct dates?

### Generate solutions
Based on the problem definition, here are some things that our program will need to do:
* Open a csv file for reading
* Extract date and price information from the first and second column on each line in the file
* Add the date and price information to a dictionary

An algorithm plan could be:
1. Create an empty dictionary
2. Open the csv file for reading
3. Use a loop to process each line in the file, for each line:
    a. Extract the date and price
    b. Add the date and price pair to the dictionary
4. Close the file


### Implement a solution
The code below is an initial attempt to complete the problem along with some code to display the contents of the dictionary. 

With your group, analyze the code and the printed output and identify the two errors in this solution. Then modify the code to resolve these errors.

In [1]:
import csv

netflix_stock_prices = {}

# open the files and extract the historical data
with open('netflix_stock_data.csv', 'r') as stock_data_file:
    stock_data_csv = csv.reader(stock_data_file)
    
    # How should we skip the first line?
    header = True
    # The price is stored as a string, how should we convert it to a float?
        
    for line in stock_data_csv:
        if header:
            header = False
        else:
            date = line[0]
            price = line[1]
            netflix_stock_prices[date] = float(price[1:])
            
        

# Let's print the contents of the dictionary using a loop
for date in netflix_stock_prices:
    print("Date: ", date, " Price: ", netflix_stock_prices[date])

Date:  11/18/2022  Price:  287.98
Date:  11/17/2022  Price:  295.28
Date:  11/16/2022  Price:  306.02
Date:  11/15/2022  Price:  310.2
Date:  11/14/2022  Price:  299.27
Date:  11/11/2022  Price:  290.13
Date:  11/10/2022  Price:  274.97
Date:  11/09/2022  Price:  254.66
Date:  11/08/2022  Price:  263.46
Date:  11/07/2022  Price:  258.6
Date:  11/04/2022  Price:  260.79
Date:  11/03/2022  Price:  269.06
Date:  11/02/2022  Price:  273.0
Date:  11/01/2022  Price:  286.75
Date:  10/31/2022  Price:  291.88
Date:  10/28/2022  Price:  295.72
Date:  10/27/2022  Price:  296.94
Date:  10/26/2022  Price:  298.62
Date:  10/25/2022  Price:  291.02
Date:  10/24/2022  Price:  282.45
Date:  10/21/2022  Price:  289.57


## Part 2 - Stock class

Now let's create a stock class that we can use to represent stocks and their prices over time.

### Define the problem
Our stock objects will have two attributes:
1. `name` - a 3 or 4 letter string representing the name of the stock, e.g. NTFX for Netflix
2. `prices` - a dictionary of date-price values

We will also give our stock class a `price_change` method which will compute and then return the change in the stock's price between two dates.

### Define test cases
A simple test case for the `price_change` method would be to check the change in price between the first and last date in our csv file.

### Generate solutions
An algorithm plan for our `price_change` method might be:
1. Lookup the price of the stock at date2 in the object's price attribute
2. Lookup the price of the stock at date1 in the object's price attribute
3. Compute and return the difference between the prices at date2 and date1

### Implement a solution
The `Stock` class is partially implemented below. With your group, complete and test the `price_change` method.

In [13]:
class Stock:
    '''
    Represents a public stock available on the stock market.
    Contains historical price information.
    '''
    
    def __init__(self,name,prices):
        self.name = name # ticker string
        self.prices = prices # dict of historical prices, indexed by date strings
        
    def __str__(self):
        return "Stock : " + self.name
    
    def price_change(self,date1,date2):
        '''
        Compute the change in the stock price
        between date2 and date1
        '''
        return self.prices[date2] - self.prices[date1]
        

# Let's create a Netflix stock object
netflix_stock = Stock('NTFX', netflix_stock_prices)

print(netflix_stock)

# check how much the price changed between November 8 and November 17
print("Price change: ", netflix_stock.price_change("11/08/2022", "11/17/2022"))

Stock : NTFX
Price change:  31.819999999999993


## Putting part 1 and 2 together

So far, we have a single stock object that we've created using data from one file.
But to make a more useful program, we'll want to have multiple stock objects created from multiple files. The function below takes a file name and stock name as input parameters and returns a `Stock` object containing the data from the file.

In [14]:
def create_stock_from_file(filename, stock_name):
    '''
    (str,str) -> Stock
    
    Given a filename and a stock name, extracts
    historical stock price information from the 
    csvfile and returns a Stock object containing
    the historical price data.
    '''
    
    stock_prices = {}
    
    with open(filename, 'r') as stock_data_file:
        stock_data_csv = csv.reader(stock_data_file)
        
        header = True
    
        for line in stock_data_csv:
            if header:
                header = False
            else:
                date = line[0]
                price = line[1]
                stock_prices[date] = float(price[1:])
                
    stock = Stock(stock_name, stock_prices)
    
    return stock


Now let's use this function to create a few more stock objects and store them in a dictionary called `available_stocks`

In [15]:
# create a number of stock objects
stock_files = {'NTFX' : 'netflix_stock_data.csv',
               'AAPL' : 'apple_stock_data.csv',
               'TSLA' : 'tesla_stock_data.csv',
               'TD'   : 'td_bank_stock_data.csv',
               'GME'  : 'gamestop_stock_data.csv'}

available_stocks = {}
for stock_name in stock_files:
    stock = create_stock_from_file(stock_files[stock_name], stock_name)
    print(stock)
    available_stocks[stock_name] = stock

Stock : NTFX
Stock : AAPL
Stock : TSLA
Stock : TD
Stock : GME


## Part 3 - Purchasing stocks

So far, we have written code to represent stocks and their historical price. Now, we want to add to program to enable us to represent stocks *owned* by individuals.

In the remainder of the exercise, we will create two more classes:
1. `Purchased_Stock`
2. `Portfolio`

We'll start with the `Purchased_Stock` class.



### Define the problem
`Purchased_Stock` objects will be used to represent purchased stocks from a single company and will have the following attributes:
1. `stock` - A `Stock` object representing the stock that was purchased
2. `number_purchased` - The number of stocks that were purchased
3. `date_purchased` - The date when the stocks were purchased

The `Purchased_Stock` class will also have the following methods:
1. `value` - computes and returns the total value of the stocks at a given date
2. `purchase_price` - Computes and returns the purchase price of the stocks on the purchase date
3. `return_on_investment` - computes and returns the "return on investment" of the stocks at a given date

return on investment (ROI) can be computed as:
$$ROI = {current\:value - initial\:cost \over initial\:cost} $$

### Define test cases
Say we bought 200 Netflix stocks on Oct. 31st.

The purchase price would be: $$Number * Value = 200 * \$291.88 = \$58376.00$$

The on Nov. 15th value would be: $62040.00

and the ROI on Nov. 15th would be: $${62040 - 58376 \over 58376} \approx 0.0628$$

### Generate solutions
`value` algorithm plan:
1. extract the price of the stock on the given date from the stock attribute
2. multiply the price by the number purchased
3. return the result

`purchase_price` algorithm plan:
1. extract the price of the stock on the date purchased from the stock attribute
2. multiply the price by the number purchased
3. return the result

`return_on_investment` algorithm plan:
1. compute the purchase price
2. compute the value at the given date
3. compute and return ROI

In [16]:
# alternative #2 with additonal class
class Purchased_Stock:
    
    def __init__(self,stock,number_purchased,date_purchased):
        self.stock = stock
        self.number_purchased = number_purchased
        self.date_purchased = date_purchased
        
    def __str__(self):
        return str(self.number_purchased) + " units of " + str(self.stock) + " purchased on " + self.date_purchased
        

    def value(self,date):
        '''
        (Purchased_Stock, str) -> float
        
        Return the total value of the stocks at a given date.
        '''
        return self.stock.prices[date] * self.number_purchased
        
    def purchase_price(self):
        '''
        (Purchased_Stock) -> float
        
        compute the amount paid to purchase the stock on the date purchased
        '''
        return self.value(self.date_purchased)
    
    
    def return_on_investment(self, date):
        '''
        (Purchased_Stock, date) -> float
        
        Compute the return on investment of the purchased stocks
        at a given date.
        '''
        current_value = self.value(date)
        initial_cost = self.value(self.date_purchased)
        roi = (current_value - initial_cost) / initial_cost
        return roi
    

In [17]:
available_stocks

{'NTFX': <__main__.Stock at 0x7fc53aa3bca0>,
 'AAPL': <__main__.Stock at 0x7fc53aa3b3a0>,
 'TSLA': <__main__.Stock at 0x7fc5398696a0>,
 'TD': <__main__.Stock at 0x7fc53994adc0>,
 'GME': <__main__.Stock at 0x7fc53994ad90>}

In [18]:
# buy 200 netflix stocks on October 31
purchased_netflix_stocks = Purchased_Stock(available_stocks['NTFX'],
                                           200,
                                           "10/31/2022")
# print the purchase price
print("Purchase price: ", purchased_netflix_stocks.purchase_price())

# check the toal value of the 200 stocks on November 15
print("Total value: ", purchased_netflix_stocks.value("11/15/2022"))

# check the return on investment on November 15
print("Return on investment: ", purchased_netflix_stocks.return_on_investment("11/15/2022"))

Purchase price:  58376.0
Total value:  62040.0
Return on investment:  0.06276552007674387


## Part 4 - Building a portfolio

Finally, let's create a `Portfolio` class that will let us represent an individual's stock portfolio that can include stocks purchased of several different companies.

`Portfolio` objects have the following attributes:
1. `purchased_stocks` - A dictionary of string-Purchased_Stock pairs where the keys are names of stocks owned
2. `total_investment` - A float representing the total amount of paid to purchase all the stocks in the portfolio

and the following methods:
1. `Purchase_Stock` - Creates a new Purchased_Stock object and adds it to the portfolio
2. `value` - Computes and returns the total value of the portfolio at a given date
3. `return_on_investment` - Computes and returns the ROI for the whole portfolio
4. `sell_stock` - Removes stock from a given company from the portfolio

The completed class is given below. Notice the similarities and differences in the `Purchased_Stock` and `Portfolio` methods.

In [19]:
class Portfolio:
    '''
    Represents a stock portfolio containing Purchased_Stocks 
    from various companies
    '''
    
    def __init__(self):
        self.purchased_stocks = {} # dict of name : Purchased_Stock
        self.total_investment = 0
        
    def purchase_stock(self,stock,number,date):
        '''
        (Portfolio, Stock, int, str) -> None
        
        Creates a new Purchased_Stock object and 
        adds it to the portfolio.
        '''
        # create the purchased stock object
        purchased_stock = Purchased_Stock(stock,number,date)
        self.purchased_stocks[stock.name] = purchased_stock
        self.total_investment += purchased_stock.purchase_price()
    
    def value(self,date):
        '''
        (Portfolio, str) -> float
        
        Computes and returns the total value of the 
        portfolio at the given date.
        '''
        total_value = 0
        for stock_name in self.purchased_stocks:
            total_value += self.purchased_stocks[stock_name].value(date)
        
        return total_value
    
    def return_on_investment(self, date):
        '''
        (Portfolio, str) -> float
        
        Computes and returns the return on investment
        for the entire portfolio at a given date.
        '''
        current_value = self.value(date)
        net_change = current_value - self.total_investment
        roi = net_change / self.total_investment
        return roi
    

    def sell_stock(self, stock_name):
        '''
        (Portfolio, str) -> None
        
        Removes stock from a given company
        from the portfolio.
        '''
        sold_stock = self.purchased_stocks.pop(stock_name)
        self.total_investment -= sold_stock.value(sold_stock.date_purchased)

## Part 4 - Building and managing a portfolio

With the `Portflio` class defined, let's write a short program to create and modify a stock porfolio.

### Define the problem
Since this is a toy example, we can just create a series of transactions and calculations for Our program should do the following:
1. Purchase 200 Netflix stocks on Oct. 31, 2022
2. Print the value of the portfolio on Nov. 7, 2022
3. Purchase 10 Apple stocks on Nov. 10, 2022
4. Purchase 49 GameStop stocks on Nov. 14, 2022
5. Compute the total return on investment on Nov. 16, 2022
6. Sell the Netflix stock and then recompute the total return on investment on Nov. 16, 2022


### Define test cases
**Portfolio value on Nov. 7**

|       Stock Name      | Number owned | Individual Purchase Cost | Total Purchase Cost | Current Individual Stock Value | Total Current Value |
|:---------------------:|:------------:|:------------------------:|:-------------------:|:------------------------------:|:-------------------:|
|          NTFX         |      200     |         \$291.88         |      \$53376.00     |            \$256.60            |      \$51720.00     |
| **Portfolio  Totals** |              |                          |    \$53376.00   |                                |    **\$51720.00**    |


**Portfolio value on Nov. 16 (before NTFX stock sale)**

|       Stock Name      | Number owned | Individual Purchase Cost | Total Purchase Cost | Current Individual Stock Value | Total Current Value |
|:---------------------:|:------------:|:------------------------:|:-------------------:|:------------------------------:|:-------------------:|
|          NTFX         |      200     |         \$291.88         |      \$58376.00     |            \$306.02            |      \$61204.00     |
|          AAPL         |      10      |         \$146.87         |      \$1468.70      |            \$148.79            |      \$1487.90      |
|          GME          |      49      |          \$26.05         |      \$1276.45      |             \$27.14            |      \$1329.86      |
| **Portfolio  Totals** |              |                          |      \$61121.15     |                                |    **\$64021.76**   |

$$ROI = {64021.76-61121.15 \over 61121.15} \approx 0.0475$$


**Portfolio value on Nov. 16 (after NTFX stock sale)**

|       Stock Name      | Number owned | Individual Purchase Cost | Total Purchase Cost | Current Individual Stock Value | Total Current Value |
|:---------------------:|:------------:|:------------------------:|:-------------------:|:------------------------------:|:-------------------:|
|          AAPL         |      10      |         \$146.87         |      \$1468.70      |            \$148.79            |      \$1487.90      |
|          GME          |      49      |          \$26.05         |      \$1276.45      |             \$27.14            |      \$1329.86      |
| **Portfolio  Totals** |              |                          |      \$2745.15     |                                |    **\$2817.76**   |

$$ROI = {2817.76-2745.15 \over 2745.15} \approx 0.0265$$

In [20]:
my_portfolio = Portfolio()

# purchase 200 Netflix stocks on Oct. 31
my_portfolio.purchase_stock(available_stocks['NTFX'], 200, "10/31/2022")

# check the value of the portfolio on Nov. 7
print("Nov 7 value: ", my_portfolio.value("11/07/2022"))

# purchase 10 Apple stocks on Nov. 10
my_portfolio.purchase_stock(available_stocks['AAPL'], 10, "11/10/2022")

# purchase 49 GameStop stocks on Nov. 14
my_portfolio.purchase_stock(available_stocks['GME'], 49, "11/14/2022")

# compute the return on investment on Nov. 16
print("Nov 16 ROI before NTFX sale: ", my_portfolio.return_on_investment("11/16/2022"))

# sell the Netflix stock on Nov. 16
my_portfolio.sell_stock('NTFX')

# re-compute the ROI on Nov. 16
print("Nov 16 ROI after NTFX sale: ", my_portfolio.return_on_investment("11/16/2022"))

Nov 7 value:  51720.00000000001
Nov 16 ROI before NTFX sale:  0.0474567314260286
Nov 16 ROI after NTFX sale:  0.02645028504817813


## BONUS: interactive portfolio manager
Below is an interactive program you can use to continue managing your portfolio. Feel free to play around and experiment with it. As you do, you will find certain conditions that cause the program to crash (hint: try buying a stock on a weekend date).
When you find these errors, think about the whole program we wrote in this problem (all classes, methods, functions) and think about how you could modify these to avoid these errors.

In [21]:
def get_user_selection():
    '''
    (None) -> str
    
    Get a user's input on how they want to manage their portfolio
    '''
    
    # Display options
    print("Please enter the number of the action you would like to perform next")
    print("\t0 - Exit")
    print("\t1 - Compute portfolio value")
    print("\t2 - Compute portfolio ROI")
    print("\t3 - Purchase stock")
    print("\t4 - Sell stock")
    
    selection = int(input("Enter your selection: "))
    
    while 0 > selection or selection > 4:
        print("Entered value: ", selection, " is not a valid option. Please select again.")
        selection = int(input("Enter your selection: "))
                
    return selection


def get_purchase_info(available_stocks):
    '''
    (dict) -> str, str, int
    
    Given a dict of available stocks, ask the user
    to indicate which stock they want to buy, the 
    number they want to buy, and the purchase date.
    '''
    
    print("The following stocks are available for purchase: ")
    i = 1
    for stock in available_stocks:
        print("\t",i,". ",stock)
        i += 1
    
    selection = input("Enter the name of the stock you would like to buy from the list above: ")
    while selection not in available_stocks:
        print("Invalid stock name, please select again.")
        selection = input("Enter the name of the stock you would like to buy from the list above: ")
    
    amount = int(input("Please enter the number of stocks you would like to purchase: "))
    while amount < 0:
        print("Please enter a positive value.")
        amount = int(input("Please enter the number of stocks you would like to purchase: "))
    
    date = input("Please enter the purchase date (MM/DD/YYYY): ")
    
    return selection, date, amount

def get_sell_info(portfolio_stocks):
    '''
    (dict) -> str
    
    Given a dict of stocks in a portfolio, ask the user
    to indicate which stock they want to sell.
    '''
    
    print("The following stocks are available to sell: ")
    i = 1
    for stock in portfolio_stocks:
        print("\t",i,". ",stock)
        i += 1
    
    selection = input("Enter the name of the stock you would like to sell from the list above: ")
    while selection not in portfolio_stocks:
        print("Invalid stock name, please select again.")
        selection = input("Enter the name of the stock you would like to sell from the list above: ")
    
    return selection

selection = -1

while selection != 0:
    
    selection = get_user_selection()
    
    if selection == 1:
        date = input("Please enter the current date (MM/DD/YYYY): ")
        print("Current portfolio value is ", my_portfolio.value(date), end="\n\n\n")
        
    elif selection == 2:
        date = input("Please enter the current date (MM/DD/YYYY): ")
        print("Current portfolio ROI is ", my_portfolio.return_on_investment(date), end="\n\n\n")
        
    elif selection == 3:
        stock_name, date, number = get_purchase_info(available_stocks)
        print("Purchasing ", number, " of ", stock_name, " stocks on ", date, end="\n\n\n")
        my_portfolio.purchase_stock(available_stocks[stock_name], number, date)
        
    elif selection == 4:
        stock_name = get_sell_info(my_portfolio.purchased_stocks)
        print("Selling ", stock_name, " from portfolio", end="\n\n\n")
        

print("Thank you!")

Please enter the number of the action you would like to perform next
	0 - Exit
	1 - Compute portfolio value
	2 - Compute portfolio ROI
	3 - Purchase stock
	4 - Sell stock
Thank you!
